app/upgrades/v1.17.0/upgrades.go

```diff
@@ -0,0 +1,61 @@
+package v1_17_0
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+
+	"github.com/provenance-io/provenance/app/keepers"
+	navs "github.com/provenance-io/provenance/app/navs"
+	"github.com/provenance-io/provenance/app/upgrades"
+	rc1 "github.com/provenance-io/provenance/app/upgrades/v1.17.0/rc1"
+	rc2 "github.com/provenance-io/provenance/app/upgrades/v1.17.0/rc2"
+	rc3 "github.com/provenance-io/provenance/app/upgrades/v1.17.0/rc3"
+	markertypes "github.com/provenance-io/provenance/x/marker/types"
+)
+
+func UpgradeStrategy(ctx sdk.Context, app upgrades.AppUpgrader, vm module.VersionMap) (module.VersionMap, error) {
+	// Migrate all the modules
+	newVM, err := upgrades.RunModuleMigrations(ctx, app, vm)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = PerformUpgrade(ctx, app.Keepers()); err != nil {
+		return nil, err
+	}
+
+	return newVM, nil
+}
+
+func PerformUpgrade(ctx sdk.Context, k *keepers.AppKeepers) (err error) {
+	if err = rc1.PerformUpgrade(ctx, k); err != nil {
```

Comment:
These three RC upgrades are all independent... does it make sense to have them all part of this same "perform upgrade" step?

Reply:
So my initial thought process was that a release should be the aggregation of all release candidates and any additional logic. I tried to extract reusable logic out of the UpgradeStrategy into PerformUpgrade so that the release could use a group of rc functions without having to call each of them independently. Would it make more sense if we renamed PerformUpgrade to something else, or if there was a separate package to share functions between release candidates and releases?

Reply:
My point was more about how the RCs lay down on testnet… they are additive but they do not run in aggregate. We also fix things in RCs that are not needed in the final upgrade… an example would be if some piece of an upgrade in one RC is found to be incorrect and it is fixed in a subsequent one. The final handler can avoid using the incorrect approach. 

From this point of view the RCs are distinct and unique updates. As a related idea to think about, when we have many networks they are likely to need their own custom logic during an update. This is why we use color names and not version numbers. The sequence of changes (thus upgrades) is independent of version numbers.

Reply:
Okay, should a release be able to use any of the functions defined by a RC such as RemoveInactiveValidatorDelegations? This would allow the release to mix and match steps between release candidates and use its own. That would be similar to how handlers within upgrades.go work now.

I didn't take multiple networks into consideration so I'll have to think about that some more. How are we planning to upgrade them? I see two approaches for upgrading them. One where they are in all sync and all chains upgrade whether or not they have logic. They all progress from 1.18.0 -> 1.18.1 -> 1.18.2. The other approach is applying upgrades to chains as they are needed. I.e. one chain is on 1.18.0 and another chain is on 1.18.x.

The UpgradeName still exists and uses the color just like we currently have. I know in github we still associate a minor version with a color so 1.18 is tourmaline. I thought ordering by version number would be easier to understand than color especially if we exceed 26 colors. "Which came first gold or green".

Reply:
It probably makes sense to have the upgrades grouped by network -- I don't see a good fit for version number based grouping like outlined above in any case ... it might also be useful to review the chain-registry repositories definition of releases and upgrades ... they list a set of releases with a list of compatible versions for each.

Something like `mainnet-tourmaline` would make it clear which network (in this case pair as testnet/mainnet are intended to be managed together).  This would also support something along the lines of `fes-epsilion` (for example).  

A longer term strategy will need to allow for other networks to use the `provenance` code base as a dependency which provides almost all of the code/functionality but allows an easily created binary release with their own compiled functionality and releases.

The SDK moved to module specific versioning where each module tracks the upgrades needed internally based on the current revision of it running on the network.  This allows modules to perform upgrades in a more self contained/known way and will likely be the best approach for a diverse set of networks.

Reply:
> So my initial thought process was that a release should be the aggregation of all release candidates and any additional logic.

That is very much not the case.

> I know in github we still associate a minor version with a color so 1.18 is tourmaline.

Nope. The `tourmaline` upgrade will do a bunch of stuff to make `mainnet` compatible with v1.18 of `provenanced`. As part of that, nodes will need to update their version of `provenanced` to `v1.18`. But `tourmaline` and `v1.18` are not the same thing. There can be upgrades without a version. There can be versions without an upgrade. An upgrade can do more than just bump the version. An upgrade doesn't have to bump the version. While we usually can get away with associating a 1-1 color and version, that isn't always the case.

---

The `RemoveInactiveValidatorDelegations` function should be defined alongside `RunModuleMigrations` since we want to remove inactive validator delegations regularly.

---

> If we aren't expecting overlap or aren't concerned with that then I can restructure it

That structure (with the colors instead of versions) is a significant improvement.

I feel like its best to not have one color reference a func in another. If we find ourselves wanting this, its probably better to copy the func to either the new color (if we don't think we'll want it again) or the common area (if we expect to want it in another future upgrade too). If one color references another, it'll almost always be the new one referencing an old function, which means that it'll break once we delete the old version.

---

> Would it make more sense if we renamed `PerformUpgrade` to something else, or if there was a separate package to share functions between release candidates and releases?

Since the `mainnet` upgrade isn't just the sum of rc upgrades, I don't find the `PerformUpgrade` function at all useful. The **only** thing useful in there would be each of the things that are done.

What would probably be even better is to focus on the `mainnet` upgrade. Put all of the sub-functions in the same file as that upgrade so that when we read it, we see exactly what actions it takes, and those actions are defined in functions below. Then, with the rc upgrades, call those functions directly too.

Basically, our first thought should usually be, "This needs to be in an upgrade, so I'll put it with the production-level upgrade." with a folloup of, "Okay, this also needs to be in our next rc." For something that shouldn't be applied to the production-level upgrade, that can go in the same file as the related rc strategy.

Either way, we still want comments explaining why something is in an rc but not full or vice versa.